File: generate-embeddings.js
import { createClient } from '@supabase/supabase-js';
import { readFileSync } from 'fs';

// Read environment variables
const envContent = readFileSync('.env', 'utf8');
const envVars = {};
envContent.split('\n').forEach(line => {
  if (line.includes('=')) {
    const [key, value] = line.split('=');
    envVars[key.trim()] = value.trim();
  }
});

const supabaseUrl = envVars.VITE_SUPABASE_URL;
const supabaseServiceKey = envVars.SUPABASE_SERVICE_ROLE_KEY;
const openAIKey = envVars.VITE_OPENAI_API_KEY || envVars.OPENAI_API_KEY;

if (!supabaseUrl || !supabaseServiceKey || !openAIKey) {
  console.error('Missing required environment variables');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseServiceKey);

async function generateEmbedding(text) {
  try {
    const response = await fetch('https://api.openai.com/v1/embeddings', {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${openAIKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        input: text.substring(0, 8000), // Limit to 8000 chars
        model: 'text-embedding-ada-002',
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      console.error('OpenAI API error:', error);
      return null;
    }

    const data = await response.json();
    return data.data[0].embedding;
  } catch (error) {
    console.error('Embedding generation error:', error);
    return null;
  }
}

async function generateEmbeddings() {
  console.log('🚀 Generating embeddings for documents...\n');

  // Get documents with content but no embeddings
  const { data: documents, error } = await supabase
    .from('documents')
    .select('id, title, content')
    .not('content', 'is', null)
    .is('vector_embedding', null)
    .is('deleted_at', null);

  if (error) {
    console.error('Error fetching documents:', error);
    return;
  }

  console.log(`Found ${documents.length} documents to process\n`);

  let processed = 0;
  let failed = 0;

  for (const doc of documents) {
    console.log(`📄 Processing: ${doc.title}`);
    
    if (!doc.content || doc.content.trim().length < 10) {
      console.log(`   ⏭️  Skipping - content too short`);
      continue;
    }

    // Generate embedding
    const embedding = await generateEmbedding(doc.content);
    
    if (embedding) {
      // Update document with embedding
      // Convert array to PostgreSQL vector format
      const vectorString = `[${embedding.join(',')}]`;
      
      const { error: updateError } = await supabase
        .from('documents')
        .update({ vector_embedding: vectorString })
        .eq('id', doc.id);

      if (updateError) {
        console.error(`   ❌ Failed to save embedding: ${updateError.message}`);
        failed++;
      } else {
        console.log(`   ✅ Embedding generated and saved`);
        processed++;
      }
    } else {
      console.log(`   ❌ Failed to generate embedding`);
      failed++;
    }

    // Rate limiting - OpenAI has rate limits
    await new Promise(resolve => setTimeout(resolve, 1000));
  }

  console.log('\n📊 Summary:');
  console.log(`✅ Successfully processed: ${processed}`);
  console.log(`❌ Failed: ${failed}`);
  console.log(`⏭️  Skipped: ${documents.length - processed - failed}`);

  if (processed > 0) {
    console.log('\n🎉 Documents are now searchable!');
    console.log('   You can now ask questions about the uploaded documents.');
  }
}

generateEmbeddings().catch(console.error);

File: scripts/generate_embeddings.py
#!/usr/bin/env python3
"""
Script to generate embeddings for existing documents in Supabase
"""
import os
import sys
import requests
import json
from typing import List, Dict
import time

# Supabase configuration - GEEN hardcoded keys meer!
SUPABASE_URL = os.environ.get("SUPABASE_URL")
SUPABASE_SERVICE_ROLE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY")

# Check of alle environment variables zijn ingesteld
if not SUPABASE_URL:
    print("Error: SUPABASE_URL environment variable is not set")
    print("Set it with: set SUPABASE_URL=your_supabase_url")
    sys.exit(1)

if not SUPABASE_SERVICE_ROLE_KEY:
    print("Error: SUPABASE_SERVICE_ROLE_KEY environment variable is not set")
    print("Set it with: set SUPABASE_SERVICE_ROLE_KEY=your_service_role_key")
    sys.exit(1)

if not OPENAI_API_KEY:
    print("Error: OPENAI_API_KEY environment variable is not set")
    print("Set it with: set OPENAI_API_KEY=your_openai_api_key")
    sys.exit(1)

headers = {
    "apikey": SUPABASE_SERVICE_ROLE_KEY,
    "Authorization": f"Bearer {SUPABASE_SERVICE_ROLE_KEY}",
    "Content-Type": "application/json"
}

def generate_embedding(text: str) -> List[float]:
    """Generate embedding using OpenAI API"""
    try:
        response = requests.post(
            "https://api.openai.com/v1/embeddings",
            headers={
                "Authorization": f"Bearer {OPENAI_API_KEY}",
                "Content-Type": "application/json"
            },
            json={
                "input": text[:8000],  # Limit text to 8000 chars
                "model": "text-embedding-ada-002"
            }
        )
        
        if response.status_code == 200:
            data = response.json()
            return data["data"][0]["embedding"]
        else:
            print(f"OpenAI API error: {response.status_code} - {response.text}")
            return None
    except Exception as e:
        print(f"Error generating embedding: {e}")
        return None

def get_documents_without_embeddings() -> List[Dict]:
    """Get all documents that have content but no embeddings"""
    response = requests.get(
        f"{SUPABASE_URL}/rest/v1/documents",
        headers=headers,
        params={
            "select": "id,title,content",
            "content": "not.is.null",
            "vector_embedding": "is.null",
            "deleted_at": "is.null"
        }
    )
    
    if response.status_code == 200:
        return response.json()
    else:
        print(f"Error fetching documents: {response.status_code} - {response.text}")
        return []

def update_document_embedding(document_id: str, embedding: List[float]) -> bool:
    """Update document with embedding"""
    response = requests.patch(
        f"{SUPABASE_URL}/rest/v1/documents",
        headers=headers,
        params={"id": f"eq.{document_id}"},
        json={"vector_embedding": embedding}
    )
    
    return response.status_code == 200

def process_documents():
    """Process all documents without embeddings"""
    documents = get_documents_without_embeddings()
    print(f"Found {len(documents)} documents to process")
    
    for i, doc in enumerate(documents):
        print(f"\nProcessing {i+1}/{len(documents)}: {doc['title']}")
        
        if not doc.get('content'):
            print("  - No content, skipping")
            continue
        
        # Generate embedding
        embedding = generate_embedding(doc['content'])
        
        if embedding:
            # Update document
            if update_document_embedding(doc['id'], embedding):
                print("  Embedding generated and saved")
            else:
                print("  Failed to save embedding")
        else:
            print("  Failed to generate embedding")
        
        # Rate limiting - OpenAI allows 3 requests per minute for free tier
        time.sleep(0.5)

if __name__ == "__main__":
    print("Starting embedding generation for documents...")
    print(f"Using Supabase URL: {SUPABASE_URL}")
    process_documents()
    print("\nDone!")